web/src/js/components/Authentication/EnterEmailForm.js

```diff
@@ -0,0 +1,217 @@
+import { replaceUrl, loginURL } from 'js/navigation/navigation'
+import EmailField from '../General/EmailField'
+import Typography from '@material-ui/core/Typography'
+
+export const PaperItem = props => (
+  <Paper
+    elevation={1}
+    style={{
+      padding: 24,
+      backgroundColor: '#FFF',
+    }}
+  >
+    <span
+      style={{
+        fontSize: 20,
+        fontWeight: 500,
+      }}
+    >
+      {props.title}
+    </span>
+    <Typography variant={'body2'} style={{ paddingTop: 24, paddingBottom: 24 }}>
+      {props.text}
+    </Typography>
+    <span
+      style={{
+        display: 'flex',
+        justifyContent: 'flex-end',
+        marginTop: 4,
+      }}
+    >
+      <Button
+        data-test-id={'enter-email-form-button'}
+        color={'primary'}
+        variant={'contained'}
+        onClick={props.buttonHandler}
+        style={{ minWidth: 96 }}
+      >
+        {props.buttonText}
+      </Button>
+    </span>
+  </Paper>
+)
+
+class EnterEmailForm extends React.Component {
+  constructor(props) {
+    super(props)
+    this.emailFieldRef = null
+    this.state = {
+      emailDuplicate: false,
+      otherError: false,
+      savingEmailInProgress: false,
+      verifyEmailSent: false,
+      inputEmail: null,
+    }
+  }
+
+  verifyEmailSent() {
+    this.setState({ verifyEmailSent: true })
+  }
+
+  updateUserFailure(error) {
+    console.log(error)
+    if (error.code === 'auth/requires-recent-login') {
+      replaceUrl(loginURL, {
+        next: '/newtab/account',
+        reauth: 'true',
+      })
+    } else if (error.code === 'auth/email-already-in-use') {
+      this.setState({ emailDuplicate: true })
+    }
+  }
+
+  submit() {
+    const email = this.emailFieldRef.getValue()
+    this.setState({
+      emailDuplicate: false,
+      otherError: false,
+      inputEmail: email,
+    })
+    updateUserEmail(email)
+      .then(() => this.verifyEmailSent())
+      .catch(this.updateUserFailure.bind(this))
+  }
+
+  resubmit() {
+    updateUserEmail(this.state.inputEmail)
+      .then(() => this.verifyEmailSent())
+      .catch(this.updateUserFailure)
+  }
+
+  onMutationCompleted(response) {
+    const { onCompleted } = this.props
+    this.setState({
+      savingEmailInProgress: false,
+    })
+    const data = response.setEmail
+
+    // Handle server-side validation errors.
+    if (!data.user) {
+      data.errors.forEach(err => {
+        // Email already exists
+        if (err.code === 'auth/email-already-in-use') {
+          this.setState({
+            emailDuplicate: true,
+          })
+          // Some other error
+        } else {
+          this.setState({
+            otherError: true,
+          })
+        }
+      })
+      return
+    }
+
+    // Go to the app.
+    onCompleted()
+  }
+
+  onMutationError(response) {
+    logger.error(response)
+    this.setState({
+      savingEmailInProgress: false,
+      otherError: true,
+    })
+  }
+
+  handleKeyPress(e) {
```

Comment:
i'm pretty sure it's possible to trigger this function twice before your set state call finishes.  You can defend against this using lodash debounce on this function